File: app/src/main/java/net/daverix/ajvm/jvm/DStoreOperation.java
package net.daverix.ajvm.jvm;


import net.daverix.ajvm.io.ByteCodeReader;

import java.io.IOException;
import java.util.Stack;

public class DStoreOperation implements ByteCodeOperation {
    @Override
    public void execute(ByteCodeReader reader,
                        int indexOfBytecode,
                        Frame currentFrame) throws IOException {
        int index = reader.readUnsignedByte();
        Object localVariable = currentFrame.pop();

        if(!(localVariable instanceof Double))
            throw new IllegalStateException("variable " + localVariable + " is not an Double");

        currentFrame.setLocalVariable(index, localVariable);
    }
}

File: app/src/main/java/net/daverix/ajvm/operation/DStoreOperation.kt
package net.daverix.ajvm.operation


import net.daverix.ajvm.Frame
import net.daverix.ajvm.ByteCodeReader

import java.io.IOException

class DStoreOperation : ByteCodeOperation {
    @Throws(IOException::class)
    override fun execute(reader: ByteCodeReader,
                         indexOfBytecode: Int,
                         currentFrame: Frame) {
        val index = reader.readUnsignedByte()
        val localVariable = currentFrame.pop()

        if (localVariable !is Double)
            throw IllegalStateException("variable $localVariable is not an Double")

        currentFrame.setLocalVariable(index, localVariable)
    }
}
